### PR TITLE
add rename for smartsync

### PR DESCRIPTION
**Pull request by earthman1**
Add smartsync aware rename method

### CODE REVIEW

File: cloudsync/smartsync.py [hunk 1]
```diff
@@ -427,3 +427,15 @@ def smart_delete_path(self, local_oid, local_path):
                 self.state.update_entry(ent, LOCAL, local_oid, path=local_path, changed=True, exists=False)
                 self.state.requestset.add(ent)
                 self.state.excludeset.discard(ent)
+
+    def smart_rename(self, side, oid, new_path) -> str:
+        """smartsync aware rename"""
+        # oid MUST exist on the specified side
+        #   then, if target exists on same side, renaming on that side will fail in the rename
+        #   if target exists only on the other side, then there will be a conflict
+        other_side = REMOTE if side == LOCAL else LOCAL
```
Comment (reed-atakama) [inline cloudsync/smartsync.py:436]: use `other_side()` function

File: cloudsync/smartsync.py [hunk 2]
```diff
@@ -427,3 +427,15 @@ def smart_delete_path(self, local_oid, local_path):
                 self.state.update_entry(ent, LOCAL, local_oid, path=local_path, changed=True, exists=False)
                 self.state.requestset.add(ent)
                 self.state.excludeset.discard(ent)
+
+    def smart_rename(self, side, oid, new_path) -> str:
+        """smartsync aware rename"""
+        # oid MUST exist on the specified side
+        #   then, if target exists on same side, renaming on that side will fail in the rename
+        #   if target exists only on the other side, then there will be a conflict
```
Comment (reed-atakama) [inline cloudsync/smartsync.py:435]: I'm a little confused on this comment. If the target exists on the other side won't we raise an error?

File: cloudsync/tests/test_cs.py [hunk 3]
```diff
@@ -3034,10 +3036,50 @@ def test_smart_delete_path_local(scs):
...
+    scs.run_until_clean(timeout)
+
+    # confirm successful regular rename
+    linfo_start = local.create(local_path1, BytesIO(contents))
+    scs.run_until_clean(timeout)
+    assert local.exists_path(local_path1)
+    assert remote.exists_path(remote_path1)
+    new_oid = scs.smart_rename(LOCAL, linfo_start.oid, local_path2)
+    scs.run_until_clean(timeout)
+    assert local.exists_path(local_path2)
```
Comment (reed-atakama) [inline cloudsync/tests/test_cs.py:3070]: could also be good to assert that the old paths don't exist

File: cloudsync/smartsync.py [hunk 4]
```diff
@@ -427,3 +427,15 @@ def smart_delete_path(self, local_oid, local_path):
                 self.state.update_entry(ent, LOCAL, local_oid, path=local_path, changed=True, exists=False)
                 self.state.requestset.add(ent)
                 self.state.excludeset.discard(ent)
+
+    def smart_rename(self, side, oid, new_path) -> str:
```
Comment (reed-atakama) [inline cloudsync/smartsync.py:431]: I'm wondering if we need an async endpoint as well for directory renames. Probably not necessary right this instant and I'm not even sure if we support directory renames for smart sync right now

Comment (earthman1) [reply cloudsync/smartsync.py:431]: Do we have any async endpoints? That's a bigger discussion. Perhaps the async code is the responsibility of the client

File: cloudsync/smartsync.py (hunk 1 above)
Comment (earthman1) [reply cloudsync/smartsync.py]: done

File: cloudsync/smartsync.py (hunk 2 above)
Comment (earthman1) [reply cloudsync/smartsync.py]: The comment reflected an earlier state of the code -- I fixed the comment

File: cloudsync/tests/test_cs.py (hunk 3 above)
Comment (earthman1) [reply cloudsync/tests/test_cs.py]: done

File: cloudsync/smartsync.py (hunk 4 above)
Comment (earthman1) [reply cloudsync/smartsync.py:439]: Do we have any async endpoints? I think the question is whether async code belongs in the library or in the client code

File: cloudsync/tests/test_cs.py [hunk 5]
```diff
@@ -3065,10 +3067,80 @@ def test_smart_delete_path_local(scs):
...
+
+
+def test_smartsync_rename_file(scs):
+    timeout = 1
+    local_parent = "/local"
+    remote_parent = "/remote"
+    local_test_folder = "/local/testfolder"
+    remote_test_folder = "/remote/testfolder"
+    remote_test_folder_contents = "/remote/testfolder/testfile"
```
Comment (howufiger7) [inline cloudsync/tests/test_cs.py:3109]: minor: not used

**Review by howufiger7 — APPROVED** [review]
LGTM

**Pull request merged**